consumer/transaction.go

```diff
@@ -443,7 +443,21 @@ func txnStartCommit(s *shard, txn *transaction) (pc.Checkpoint, error) {
 	var waitFor = s.ajc.PendingExcept(s.recovery.log)
 
 	trace.Log(s.ctx, "store.StartCommit", s.resolved.fqn)
-	txn.commitBarrier = s.store.StartCommit(s, txn.checkpoint, waitFor)
+	var barrier = s.store.StartCommit(s, txn.checkpoint, waitFor)
+
+	// If StartCommit returned an OpFuture which is a pre-resolved error,
+	// handle it as an immediate and synchronous error. This avoids raced
+	// calls to Application.ConsumeMessage when the Application has already
+	// failed.
+	select {
+	case <-barrier.Done():
+		if barrier.Err() != nil {
+			return pc.Checkpoint{}, fmt.Errorf("store.StartCommit: %w", barrier.Err())
+		}
+	default:
+	}
+
+	txn.commitBarrier = barrier
```

Comment:
The `txn.commitBarrier` now won't be set if we're synchronously returning the error. Are all the other accesses of `commitBarrier` OK with that? Should it clear the previous value? I looked through the file and didn't see any problems, but just wanted to double check.

Reply:
yep 👍